File: distros/J/JavonetPerlSdk/lib/Javonet/Binaries/Ruby/MacOs/X64/core/handler/invoke_instance_method_handler.rb
require_relative 'abstract_command_handler'

class InvokeInstanceMethodHandler < AbstractCommandHandler
  def initialize
    @required_parameters_count = 2
  end

  def process(command)
    return invoke_instance_method(command)
  end

  def invoke_instance_method(command)
    begin
      if command.payload.length < @required_parameters_count
        raise ArgumentError.new "InvokeInstanceMethod parameters mismatch"
      end
      if command.payload.length > 2
        arguments = command.payload[2..]
        return command.payload[0].send(command.payload[1], *arguments)
      else
        return command.payload[0].send(command.payload[1])
      end
    rescue NoMethodError
      methods = command.payload[0].methods
      message = "Method #{command.payload[1]} not found in object of class #{command.payload[0].class.name}. Available methods:\n"
      methods.each { |method_iter| message += "#{method_iter}\n" }
      raise Exception, message
    rescue Exception => e
      return e
    end
  end
end

File: distros/J/JavonetPerlSdk/lib/Javonet/Binaries/Nodejs/core/handler/InvokeInstanceMethodHandler.js
const AbstractHandler = require("./AbstractHandler");

class InvokeInstanceMethodHandler extends AbstractHandler {
  requiredParametersCount = 2;

  constructor() {
    super();
  }

  process(command) {
    try {
      if (command.payload.length < this.requiredParametersCount) {
        throw new Error("Invoke Instance Method parameters mismatch");
      }
      const { payload } = command;
      let instance = payload[0];
      let methodName = payload[1];
      let args = payload.slice(2);
      let method = instance[methodName];
      if (typeof method === "undefined") {
        let methods = Object.getOwnPropertyNames(instance.__proto__).filter(
          function (property) {
            return typeof instance.__proto__[property] === "function";
          }
        );
        let message = `Method ${methodName} not found in object. Available methods:\n`;
        methods.forEach((methodIter) => {
          message += `${methodIter}\n`;
        });
        throw new Error(message);
      } else {
        return Reflect.apply(method, instance, args);
      }
    } catch (error) {
      throw this.process_stack_trace(error, this.constructor.name);
    }
  }
}

module.exports = new InvokeInstanceMethodHandler();
